doc/specs/#885 - winrt Terminal Settings.md

```diff
@@ -0,0 +1,205 @@
+
+TerminalApp will construct and reference an `AppSettings settings` using a shared smart pointer as follows:
+- TerminalApp will have a global reference to "defaults.json" and "settings.json" filepaths as `constexpr`
+- construct an `AppSettings` using the default constructor
+- `settings.LayerSettings("defaults.json")`
+- `settings.LayerSettings("settings.json")`
+
+**NOTE:** This model allows us to layer even more settings files on top of the existing Terminal Settings
+ Model, if so desired. This could be helpful when importing additional settings files from an external location
+ such as a marketplace.
+
+When TerminalApp detects a change to "settings.json", it'll repeat the steps above. We could cache the result from
+ `settings.LayerSettings("defaults.json")` to improve performance.
+
+Additionally, TerminalApp would reference `settings` for the following functionality...
+- `Profiles` to populate the dropdown menu
+- `Keybindings` to detect active key bindings and which actions they run
+- `Commands` to populate the Command Palette
+- `Warnings` to show a serialization warning, if any
+
+#### TerminalControl: Acquiring and Applying the Settings
+
+At the time of writing this spec, TerminalApp constructs `TerminalControl.TerminalSettings` WinRT objects
+ to expose `IControlSettings` and `ICoreSettings` to any hosted terminals. In moving `IControlSettings`
+ and `ICoreSettings` down to the TerminalControl layer, TerminalApp can now have better control over
+ how to expose relevant settings to a TerminalControl instance.
+
+`TerminalSettings` (which implements `IControlSettings` and `ICoreSettings`) will be moved to TerminalApp and act as a bridge that queries `AppSettings`. Thus,
+ when TermControl requests `IControlSettings` data, `TerminalSettings` responds by extracting the
+ relevant information from `AppSettings`. TerminalCore operates the same way with `ICoreSettings`.
+
+#### Settings UI: Modifying and Applying the Settings
+
+The Settings UI will also have a reference to the `AppSettings settings` from TerminalApp using a shared smart pointer
+as `settingsSource`. When the Settings UI is opened up, the Settings UI will also have its own `AppSettings settingsClone`
+that is a clone of TerminalApp's `AppSettings`.
+```c++
+settingsClone = settingsSource.Clone()
+```
+
+As the user navigates the Settings UI, the relevant contents of `settingsSource` will be retrieved and presented.
+As the user makes changes to the Settings UI, XAML will hold on to the changes that are made.
+When the user saves/applies the changes in the XAML, `settingsClone` will be updated to reflect those changes,
+then `settingsClone.ApplyChanges("settings.json", settingsSource)` is called. This compares the changes between
+`settingsClone` and `settingsSource`, then injects the changes (if any) to `settings.json`.
+
+As mentioned earlier, TerminalApp detects a change to "settings.json" to update its `AppSettings`. 

```

Comment:
So wait

1. settings is cloned
2. the UI reads from the base settings
3. the UI saves into the clone
4. We take the clone, go to the base settings, and tell the base settings to "apply" the clone to .. the file
5. _the base settings object is not updated (!)_
6. the base settings object writes the changes to the file
7. TerminalApp goes around and recreates the base settings object from the file?



what is the point of having an `ApplySettings` API if all it does is _diff my contents against someone else's contents and write that out to a file_?

Reply:
So, my main concern is what happens if somebody has `settings.json` and the Settings UI open at the same time. They need to keep in sync (along with the TermApp settings object too).

Since we already have a mechanism for detecting/reloading changes from `settings.json` in TermApp, I figured we could leverage that. The Settings UI should just be a view of what we are using in TermApp, then directly modifying the JSON. Since JSON is master, everything gets updated from touching the JSON.

Regarding step 5 (the base settings object is not updated). We _could_ update it. But then we're completely overwriting it to be the same thing when TermApp listens to a change in settings.json.

Reply:
What happens when you have two text editors open to settings.json? Best case scenario, one tells you "this file changed. want to reload it?".

This is a case of "play stupid games, win stupid prizes".

Reply:
The settings UI should be a view over the clone of settings, not the live copy, and it's not just editing the JSON directly

Reply:
> The settings UI should be a view over the clone of settings, not the live copy, and it's not just editing the JSON directly

Ok. I've been convinced. :)
We do it this way. If somebody is editing the settings via both a text editor and the Settings UI, when you save/apply in whichever app you're using, you are _overwriting_ not _merging_.

I guess also, if users feel that this is a more common problem, we can handle it then.

Reply:
urgh. Well, now I'm thinking about the longer-term. Since some settings won't be in the Settings UI, it's not that unexpected for users to still edit the JSON. Like, I could see myself editing the JSON _and_ the Settings UI at the same time (but idk how much of that is because I am so close to the project).

Leaving this unresolved. We have a meeting scheduled for tomorrow so I'll be sure to bring this up for discussion then :)

Reply:
This section has been moved to future considerations.

Also talked about it offline. Sounds like your main concern was the concept of "merging" which we're not doing.